File: python/src/toolnexus/acp.py
"""ACP (Agent Client Protocol) as a model source — ADR 0031, issue #96.

ACP is to *agents* what MCP is to *tools*: JSON-RPC 2.0, one object per line, over a
child process's stdin/stdout — the exact framing MCP local stdio already uses
(``mcp_source.py``). `devin acp`, Gemini CLI and Zed's agents speak it. This module
ships ACP as the *smallest* possible framing per ADR 0031: a ``generate(request) ->
dict`` function with the shape ``create_in_process_client(generate=...)`` already
wants (``client.py`` around ``InProcessTransport``/``create_in_process_client``), so
the tool-calling loop, skills, MCP tools, adapters and sub-agents are untouched. ACP
is a model source, not a new tool source and not a new client.

The hard part isn't the wire protocol (mechanical: ``initialize`` -> ``session/new``
-> ``session/prompt``, demultiplexed by JSON-RPC id because ``session/update``
notifications interleave with responses). It's two behaviors the spike at
``spikes/acp/`` proved matter:

* **Only ``agent_message_chunk`` forms the reply.** ``agent_thought_chunk`` and tool
  narration must be dropped, or they corrupt structured output the host expects to
  parse.
* **A permission request is answered, never awaited.** ``session/request_permission``
  left unanswered hangs a turn forever, even in an agent's "bypass" mode — so this
  client answers inline, from the same thread that demultiplexes the wire, with the
  first option whose ``kind`` starts with ``allow``.

toolnexus assembles a *complete* request every turn (the full message array), but an
ACP session is *stateful* — it already holds the transcript. Sending the whole thing
every turn makes a session accumulate near-duplicate history, and ADR 0031's reporter
observed the agent answering a stale copy. The mitigation (kept as the *library's*
responsibility, not left to the caller) is an explicit marker naming the current turn
as superseding everything earlier — see ``_SUPERSEDES_MARKER`` and
``_render_prompt``.

Sync-bridge design (the reason this file doesn't need ``asyncio``, and why that's the
*right* choice here, not a workaround): ``InProcessTransport.post`` (``client.py``)
rejects an awaitable answer from ``generate`` — it MUST be a plain synchronous
callable, because it runs on the client's own worker thread. This client honors that
by construction rather than fighting it: ``load_acp`` starts one background
**reader thread** that owns the child's stdout loop end to end — parsing
newline-delimited JSON-RPC, resolving replies to *our* requests by id into a
``queue.Queue`` slot, routing ``session/update`` notifications into whichever turn is
currently active, and answering ``session/request_permission`` inline, all from that
one thread. The public ``ACPClient.generate`` stays a plain synchronous function: it
writes the ``session/prompt`` request, then blocks on ``queue.Queue.get(timeout=...)``
until the reader thread delivers the turn's reply. Nothing async ever crosses the
``generate`` boundary, so ``client.py``'s ``inspect.isawaitable`` guard never fires —
and the session is still warm and asynchronous *underneath*. This mirrors
``spikes/portability/SPIKE.md``'s own recommendation for Python specifically
("a background reader thread demultiplexing JSON-RPC notifications from responses
while `generate` blocks synchronously waiting for its turn's answer — doable with
`threading` + a `queue.Queue`, no new dependency") — the seam is widened nowhere;
``client.py`` is not touched.

Turns on one session are one conversation, so one ``threading.Lock`` held for the
duration of a single ``generate()`` call serialises them — idiomatic, and the
simplest thing that satisfies "one session = one conversation, concurrent prompts
would interleave into one transcript."
"""
from __future__ import annotations

import json
import os
import queue
import subprocess
import threading
from dataclasses import dataclass, field
from typing import Any, Callable, Optional

__all__ = [
    "ACPOptions",
    "ACPClient",
    "ACPError",
    "ACPPermissionTimeoutError",
    "load_acp",
]

# Marker convention shared with the Go spike (spikes/acp/fakeagent/main.go,
# spikes/acp/client/client.go) for cross-port consistency, not because the exact
# wording is part of any protocol contract — any explicit, unambiguous marker works.
_SUPERSEDES_MARKER = "SUPERSEDES-ALL-PRIOR"

# Bound for the handshake calls (initialize / session/new / session/set_mode). These
# are cheap control-plane calls even against a slow-starting real agent binary, so a
# generous fixed timeout (independent of ACPOptions.permission_timeout, which bounds
# a *turn*) is appropriate here.
_HANDSHAKE_TIMEOUT = 30.0


class ACPError(RuntimeError):
    """An ACP protocol-level failure — a JSON-RPC error reply, or a call that timed
    out waiting on the wire."""


class ACPPermissionTimeoutError(ACPError):
    """A turn did not receive its ``session/prompt`` reply within
    ``ACPOptions.permission_timeout``.

    This is the safety net for the trap ADR 0031 names explicitly: an unanswered
    ``session/request_permission`` hangs a turn forever. This client always answers
    permission requests itself (see module docstring), so in the normal path this
    timeout never fires — it exists only so a broken or non-conformant agent cannot
    hang a caller (or a test suite) indefinitely.
    """


@dataclass
class ACPOptions:
    """Configuration for one ACP session.

    ``cwd`` defaults to the current process's absolute working directory — real ACP
    agents (``devin acp`` confirmed live in the spike) reject ``session/new`` with
    ``-32602 Invalid params`` without an *absolute* ``cwd``, so this is resolved to an
    absolute path unconditionally, never left relative.
    """

    command: str
    args: list[str] = field(default_factory=list)
    cwd: Optional[str] = None
    env: Optional[dict[str, str]] = None
    protocol_version: int = 1
    permission_timeout: float = 5.0
    mode: Optional[str] = None


def _content_to_text(content: Any) -> str:
    """Flatten a message's ``content`` — a plain string, or a §1B content-part list
    — to text. Non-text parts (image/audio/file) are silently skipped: ACP's prompt
    turn is text-only in this client, matching the ``session/prompt`` shape used
    throughout the spike and the ADR."""
    if content is None:
        return ""
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        parts: list[str] = []
        for item in content:
            if isinstance(item, dict):
                text = item.get("text")
                if text is None and item.get("type") == "text":
                    text = item.get("value")
                if text is not None:
                    parts.append(str(text))
            elif isinstance(item, str):
                parts.append(item)
        return "".join(parts)
    return str(content)


def _render_prompt(request: dict[str, Any]) -> str:
    """Render the full accumulated transcript as one prompt string, per ADR 0031's
    chosen default (full request every turn, not a delta), and append the
    supersedes marker naming the latest user turn.

    This is deliberately the library's job, not the caller's — ADR 0031 rejected
    leaving the marker to be bolted on by hand ("ship the supersedes marker as part
    of the default `Generate`'s prompt assembly (not left to the caller)",
    ``spikes/acp/SPIKE.md``).
    """
    messages = request.get("messages") or []
    lines: list[str] = []
    last_user_text = ""
    for message in messages:
        if not isinstance(message, dict):
            continue
        role = message.get("role", "")
        text = _content_to_text(message.get("content"))
        if text:
            lines.append(f"{role}: {text}")
            if role == "user":
                last_user_text = text
    if not last_user_text and lines:
        last_user_text = lines[-1]
    lines.append(f"{_SUPERSEDES_MARKER}: {last_user_text}")
    return "\n".join(lines)


class ACPClient:
    """One warm ACP session: one child process, one ``sessionId``, alive across many
    ``generate()`` turns. Construct via :func:`load_acp`, not directly.
    """

    def __init__(self, opts: ACPOptions, proc: "subprocess.Popen[str]") -> None:
        self._opts = opts
        self._proc = proc

        self._write_lock = threading.Lock()
        self._state_lock = threading.Lock()
        self._turn_lock = threading.Lock()  # serialises generate() calls (one session = one conversation)
        self._close_lock = threading.Lock()

        self._id_counter = 0
        self._pending: dict[str, "queue.Queue[dict[str, Any]]"] = {}
        self._session_id: Optional[str] = None
        self._closed = False

        self._accum: list[str] = []

        self._reader = threading.Thread(target=self._read_loop, name="acp-reader", daemon=True)
        self._reader.start()

    # -- wire plumbing -------------------------------------------------- #

    def _next_id(self) -> str:
        with self._state_lock:
            self._id_counter += 1
            return f"c-{self._id_counter}"

    def _write(self, obj: dict[str, Any]) -> None:
        line = json.dumps(obj) + "\n"
        with self._write_lock:
            stdin = self._proc.stdin
            if stdin is None:
                raise ACPError("acp: child process has no stdin")
            try:
                stdin.write(line)
                stdin.flush()
            except (BrokenPipeError, ValueError, OSError) as exc:
                raise ACPError(f"acp: failed to write to child process: {exc}") from exc

    def _call(self, method: str, params: dict[str, Any], timeout: float = _HANDSHAKE_TIMEOUT) -> dict[str, Any]:
        req_id = self._next_id()
        reply_q: "queue.Queue[dict[str, Any]]" = queue.Queue(maxsize=1)
        with self._state_lock:
            self._pending[req_id] = reply_q
        self._write({"jsonrpc": "2.0", "id": req_id, "method": method, "params": params})
        try:
            msg = reply_q.get(timeout=timeout)
        except queue.Empty:
            with self._state_lock:
                self._pending.pop(req_id, None)
            raise ACPError(f"acp: {method} timed out after {timeout}s") from None
        error = msg.get("error")
        if error:
            raise ACPError(f"acp error {error.get('code')}: {error.get('message')}")
        return msg.get("result") or {}

    def _read_loop(self) -> None:
        stdout = self._proc.stdout
        if stdout is None:
            return
        try:
            for raw_line in stdout:
                line = raw_line.strip()
                if not line:
                    continue
                try:
                    msg = json.loads(line)
                except json.JSONDecodeError:
                    continue
                self._handle_message(msg)
        except (ValueError, OSError):
            pass
        finally:
            self._drain_pending_on_eof()

    def _drain_pending_on_eof(self) -> None:
        with self._state_lock:
            pending = list(self._pending.items())
            self._pending.clear()
        eof_error = {"error": {"code": -1, "message": "acp: child process closed the connection"}}
        for _req_id, q in pending:
            try:
                q.put_nowait(eof_error)
            except queue.Full:  # pragma: no cover - maxsize=1, always empty here
                pass

    def _handle_message(self, msg: dict[str, Any]) -> None:
        method = msg.get("method")
        has_id = "id" in msg and msg.get("id") is not None

        if not method and has_id:
            # A reply to one of OUR requests (initialize / session/new /
            # session/set_mode / session/prompt).
            req_id = msg["id"]
            with self._state_lock:
                q = self._pending.pop(req_id, None)
            if q is not None:
                try:
                    q.put_nowait(msg)
                except queue.Full:  # pragma: no cover - maxsize=1, always empty here
                    pass
            return

        if method == "session/update":
            self._handle_update(msg.get("params") or {})
            return

        if method == "session/request_permission" and has_id:
            self._answer_permission(msg)
            return

        # Any other server-initiated request/notification is outside this client's
        # scope (e.g. fs/read_text_file — this client declares fs capabilities as
        # false, so a conformant agent won't send one). Reply with an empty result
        # to any unknown *request* so a strict peer doesn't itself hang on us.
        if has_id:
            self._write({"jsonrpc": "2.0", "id": msg["id"], "result": {}})

    def _handle_update(self, params: dict[str, Any]) -> None:
        update = params.get("update") or {}
        if update.get("sessionUpdate") != "agent_message_chunk":
            # agent_thought_chunk and tool_call/tool_call_update narration are
            # dropped here, on purpose — mixing them into the reply is exactly the
            # corruption ADR 0031 gate item 3 proves against structured output.
            return
        content = update.get("content") or {}
        text = content.get("text")
        if text:
            with self._state_lock:
                self._accum.append(str(text))

    def _answer_permission(self, request: dict[str, Any]) -> None:
        params = request.get("params") or {}
        options = params.get("options") or []
        chosen: Optional[str] = None
        for option in options:
            kind = (option or {}).get("kind") or ""
            if kind.startswith("allow"):
                chosen = option.get("optionId")
                break
        if chosen is not None:
            result = {"outcome": {"outcome": "selected", "optionId": chosen}}
        else:
            result = {"outcome": {"outcome": "cancelled"}}
        self._write({"jsonrpc": "2.0", "id": request["id"], "result": result})

    # -- handshake -------------------------------------------------------- #

    def _connect(self) -> None:
        self._call(
            "initialize",
            {
                "protocolVersion": self._opts.protocol_version,
                "clientCapabilities": {"fs": {"readTextFile": False, "writeTextFile": False}},
            },
        )
        cwd = self._opts.cwd or os.getcwd()
        if not os.path.isabs(cwd):
            cwd = os.path.abspath(cwd)
        result = self._call("session/new", {"cwd": cwd, "mcpServers": []})
        self._session_id = result.get("sessionId")
        if self._opts.mode:
            self._call("session/set_mode", {"sessionId": self._session_id, "modeId": self._opts.mode})

    # -- public API --------------------------------------------------------- #

    def generate(self, request: dict[str, Any]) -> dict[str, Any]:
        """``Callable[[dict], dict]`` — the exact shape ``create_in_process_client``
        wants. Sends exactly one ``session/prompt`` on the warm session, accumulates
        only ``agent_message_chunk`` text, and blocks (synchronously) until the
        reader thread delivers the reply or ``permission_timeout`` elapses.
        """
        with self._turn_lock:
            prompt_text = _render_prompt(request)
            with self._state_lock:
                self._accum = []

            req_id = self._next_id()
            reply_q: "queue.Queue[dict[str, Any]]" = queue.Queue(maxsize=1)
            with self._state_lock:
                self._pending[req_id] = reply_q

            self._write(
                {
                    "jsonrpc": "2.0",
                    "id": req_id,
                    "method": "session/prompt",
                    "params": {
                        "sessionId": self._session_id,
                        "prompt": [{"type": "text", "text": prompt_text}],
                    },
                }
            )

            try:
                msg = reply_q.get(timeout=self._opts.permission_timeout)
            except queue.Empty:
                with self._state_lock:
                    self._pending.pop(req_id, None)
                raise ACPPermissionTimeoutError(
                    "acp: turn hung waiting on a reply (possibly an unanswered "
                    f"session/request_permission) after {self._opts.permission_timeout}s"
                ) from None

            error = msg.get("error")
            if error:
                raise ACPError(f"acp error {error.get('code')}: {error.get('message')}")

            with self._state_lock:
                content = "".join(self._accum)
            return {"content": content}

    def is_alive(self) -> bool:
        """True while the child process is still running."""
        return self._proc.poll() is None

    def close(self) -> None:
        """Terminate the child process. Idempotent — safe to call more than once,
        and safe to call regardless of any in-flight turn's outcome (process
        lifetime is independent of any one turn's cancellation, per ADR 0031)."""
        with self._close_lock:
            if self._closed:
                return
            self._closed = True

        stdin = self._proc.stdin
        if stdin is not None:
            try:
                stdin.close()
            except (BrokenPipeError, ValueError, OSError):
                pass
        try:
            self._proc.terminate()
        except OSError:
            pass
        try:
            self._proc.wait(timeout=5)
        except subprocess.TimeoutExpired:
            try:
                self._proc.kill()
                self._proc.wait(timeout=5)
            except (OSError, subprocess.TimeoutExpired):  # pragma: no cover - best effort
                pass

        self._drain_pending_on_eof()


def load_acp(opts: ACPOptions) -> ACPClient:
    """Spawn the ACP agent as a child process, complete the handshake
    (``initialize`` -> ``session/new`` -> optional ``session/set_mode``), and return
    a warm :class:`ACPClient` whose bound ``generate`` method is ready to hand
    straight to ``create_in_process_client(generate=client.generate, ...)``.
    """
    env = None
    if opts.env is not None:
        env = {**os.environ, **opts.env}

    cwd = opts.cwd or os.getcwd()
    if not os.path.isabs(cwd):
        cwd = os.path.abspath(cwd)

    proc = subprocess.Popen(
        [opts.command, *opts.args],
        stdin=subprocess.PIPE,
        stdout=subprocess.PIPE,
        stderr=subprocess.DEVNULL,
        cwd=cwd,
        env=env,
        text=True,
        bufsize=1,
    )

    client = ACPClient(opts, proc)
    try:
        client._connect()
    except Exception:
        client.close()
        raise
    return client

File: elixir/lib/toolnexus/acp.ex
defmodule Toolnexus.Acp do
  @moduledoc """
  ACP (Agent Client Protocol) model source — issue #96, ADR 0025, proposal
  `openspec/changes/add-acp-model-source`.

  Connects to a running agent (`devin acp`, Gemini CLI, Zed's agents, ...) over
  a child process's stdin/stdout, speaking JSON-RPC 2.0, one object per line —
  the SAME framing MCP local stdio already uses (SPEC §2), so this module
  reuses `Toolnexus.Mcp.Transport.Stdio` to spawn the child rather than
  hand-rolling a second port-spawning path.

  A **warm session** (ADR 0025 — "the warm session is the feature"): the
  process is spawned once via `connect/2`, `initialize` and `session/new`
  happen once, and every subsequent turn is one `session/prompt` against that
  same session. Concretely:

    * Only `agent_message_chunk` `session/update` notifications are
      accumulated into the reply — `agent_thought_chunk` and tool-call
      narration are dropped, or they wrap prose around structured output and
      break JSON parsing outright.
    * Responses to OUR calls and `session/update` notifications interleave on
      the same stream, so replies are demultiplexed by JSON-RPC id.
    * `session/request_permission` is answered INLINE, from the read loop,
      with the first `allow`-kind option, the instant it arrives — never
      exposed to the caller and never awaited. An unanswered permission
      request hangs the turn forever (even in bypass mode); answering
      immediately is what this module is FOR.
    * Turns are serialised: one ACP session is one conversation, so a second
      concurrent `generate/1` call is queued rather than interleaved into the
      same transcript.
    * The child process's lifetime is independent of any one turn — a failed
      or errored turn does not kill the session — and `close/1` is
      idempotent.
    * Every turn sends the FULL assembled request (every message, every
      turn — matching every other toolnexus model source, which is
      stateless by default) PLUS an explicit supersedes marker built by this
      library (not left to the caller), naming the current turn's content —
      the mitigation `spikes/acp/SPIKE.md` gate 1 proved against a stateful
      session that otherwise answers a near-duplicate, stale prompt.

  `generate/1` returns the exact `(map() -> map())` shape
  `Toolnexus.Client.create_in_process/1` and `Toolnexus.Agents.Runtime`'s
  `:in_process` option already accept, so the tool-calling loop, skills, MCP
  and sub-agents are completely unmodified — ACP is a model source, not a new
  tool source and not a new client. See `spikes/acp/SPIKE.md` for the
  reference Go client this ports, and `docs/adr/0025-acp-a-warm-session-is-the-feature.md`.
  """

  use GenServer, restart: :temporary, shutdown: 10_000

  alias Toolnexus.Mcp.Transport.Stdio

  @default_timeout 30_000
  @supersedes_marker "SUPERSEDES-ALL-PRIOR:"

  # --------------------------------------------------------------------------
  # client API
  # --------------------------------------------------------------------------

  @doc """
  Spawn `command` (a list, e.g. `["devin", "acp"]`) and complete the ACP
  handshake: `initialize` then `session/new`.

  Options:

    * `:cwd` — **absolute** working directory advertised to the agent.
      Defaults to `File.cwd!()`. A real `devin acp` rejects `session/new`
      with `-32602 Invalid params` on a relative or missing `cwd`.
    * `:mcp_servers` — the `mcpServers` array on `session/new` (default `[]`).
      Real `devin acp` requires the key to be present, even empty.
    * `:environment` / `:env` — extra env vars for the child.
    * `:timeout` — per-RPC-call timeout in ms (default 30000), covering
      `initialize` and `session/new` during connect, and each
      `session/prompt` reply.

  Returns `{:ok, pid}` or `{:error, reason}`.
  """
  @spec connect([String.t()], keyword()) :: {:ok, pid()} | {:error, term()}
  def connect(command, opts \\ []) when is_list(command) do
    GenServer.start_link(__MODULE__, {command, opts})
  end

  @doc """
  Build a semantic `generate` function — `(map() -> map())` — backed by this
  warm ACP session, in the exact shape `Toolnexus.Client.create_in_process/1`
  and `Toolnexus.Agents.Runtime`'s `:in_process` option accept.

      {:ok, acp} = Toolnexus.Acp.connect(["devin", "acp"])
      client = Toolnexus.Client.create_in_process(model: "devin", generate: Toolnexus.Acp.generate(acp))

  Raises on an ACP-level failure (a closed session, a timed-out call, an RPC
  error) — the same failure shape any other `generate` function raising
  surfaces to the caller.
  """
  @spec generate(pid()) :: (map() -> map())
  def generate(pid) do
    fn req ->
      text = assemble_prompt_text(Map.get(req, :messages) || Map.get(req, "messages") || [])

      case prompt(pid, text) do
        {:ok, answer} ->
          %{content: answer}

        {:error, reason} ->
          raise "toolnexus: ACP generate failed: #{inspect(reason)}"
      end
    end
  end

  @doc """
  Send one turn's FULL prompt text and wait for the accumulated
  `agent_message_chunk` reply. Turns on the same session are serialised.
  Exposed as the lower-level primitive `generate/1` builds on; most callers
  should use `generate/1`.
  """
  @spec prompt(pid(), String.t(), timeout()) :: {:ok, String.t()} | {:error, term()}
  def prompt(pid, text, timeout \\ :infinity) do
    GenServer.call(pid, {:prompt, text}, timeout)
  catch
    :exit, reason -> {:error, {:acp_exit, exit_brief(reason)}}
  end

  @doc "Close the session. Idempotent — safe to call more than once."
  @spec close(pid()) :: :ok
  def close(pid) do
    if Process.alive?(pid) do
      GenServer.call(pid, :close, 10_000)
      :ok
    else
      :ok
    end
  catch
    :exit, _ -> :ok
  end

  defp exit_brief({reason, _call}), do: reason
  defp exit_brief(reason), do: reason

  # --------------------------------------------------------------------------
  # prompt assembly — full request every turn + the library-built supersedes
  # marker (ADR 0025's proposed default; spikes/acp/SPIKE.md gate 1).
  # --------------------------------------------------------------------------

  @doc false
  def assemble_prompt_text(messages) do
    transcript =
      messages
      |> Enum.map(fn m -> "#{msg_role(m)}: #{msg_text(m)}" end)
      |> Enum.join("\n")

    latest = messages |> List.last() |> then(fn m -> (m && msg_text(m)) || "" end)
    marker = @supersedes_marker <> " " <> latest

    case transcript do
      "" -> marker
      t -> t <> "\n" <> marker
    end
  end

  defp msg_role(m), do: to_string(Map.get(m, "role") || Map.get(m, :role) || "user")

  defp msg_text(m) do
    content = Map.get(m, "content") || Map.get(m, :content)

    cond do
      is_binary(content) -> content
      is_list(content) -> content |> Enum.map(&extract_part_text/1) |> Enum.join("")
      true -> ""
    end
  end

  defp extract_part_text(%{"text" => t}), do: to_string(t)
  defp extract_part_text(%{text: t}), do: to_string(t)
  defp extract_part_text(_), do: ""

  # --------------------------------------------------------------------------
  # GenServer
  # --------------------------------------------------------------------------

  @impl true
  def init({command, opts}) do
    cfg = %{
      "command" => command,
      "cwd" => Keyword.get(opts, :cwd),
      "environment" => Keyword.get(opts, :environment) || Keyword.get(opts, :env) || %{}
    }

    timeout = Keyword.get(opts, :timeout, @default_timeout)

    case Stdio.open(cfg) do
      {:ok, stdio} ->
        case handshake(stdio, opts, timeout) do
          {:ok, session_id} ->
            {:ok,
             %{
               stdio: stdio,
               buffer: "",
               next_id: 2,
               session_id: session_id,
               active_from: nil,
               active_id: nil,
               chunks: [],
               queue: :queue.new(),
               timeout: timeout,
               closed: false
             }}

          {:error, reason} ->
            Stdio.close(stdio)
            {:stop, reason}
        end

      {:error, reason} ->
        {:stop, reason}
    end
  end

  # The handshake (`initialize` then `session/new`) runs synchronously inside
  # init/1, draining raw Port messages directly — nothing else can be talking
  # to this process yet (the GenServer loop has not started dispatching), so a
  # small blocking read loop here is safe and keeps connect/2 simple: by the
  # time `start_link` returns, the session genuinely exists.
  defp handshake(stdio, opts, timeout) do
    init_id = "c-1"

    send_req(stdio, init_id, "initialize", %{
      "protocolVersion" => 1,
      "clientCapabilities" => %{"fs" => %{"readTextFile" => false, "writeTextFile" => false}}
    })

    with {:ok, _} <- await_handshake_reply(stdio, init_id, "", timeout) do
      new_id = "c-2"
      cwd = Keyword.get(opts, :cwd) || File.cwd!()
      mcp_servers = Keyword.get(opts, :mcp_servers, [])

      send_req(stdio, new_id, "session/new", %{"cwd" => cwd, "mcpServers" => mcp_servers})

      case await_handshake_reply(stdio, new_id, "", timeout) do
        {:ok, result} -> {:ok, result["sessionId"]}
        {:error, _} = err -> err
      end
    end
  end

  defp await_handshake_reply(%Stdio{port: port} = stdio, want_id, buffer, timeout) do
    receive do
      {^port, {:data, chunk}} ->
        {lines, rest} = Stdio.split_lines(buffer <> chunk)

        case find_reply(lines, want_id) do
          {:found, result} -> result
          :not_found -> await_handshake_reply(stdio, want_id, rest, timeout)
        end

      {^port, {:exit_status, code}} ->
        {:error, {:acp_process_exited, code}}
    after
      timeout -> {:error, :timeout}
    end
  end

  defp find_reply([], _id), do: :not_found

  defp find_reply([line | rest], id) do
    case Jason.decode(line) do
      {:ok, %{"id" => ^id, "result" => result}} -> {:found, {:ok, result}}
      {:ok, %{"id" => ^id, "error" => err}} -> {:found, {:error, {:acp_error, err}}}
      _ -> find_reply(rest, id)
    end
  end

  defp send_req(stdio, id, method, params) do
    Stdio.send_msg(stdio, %{"jsonrpc" => "2.0", "id" => id, "method" => method, "params" => params})
  end

  # -- turn serialisation -----------------------------------------------------

  @impl true
  def handle_call({:prompt, _text}, _from, %{closed: true} = state) do
    {:reply, {:error, :closed}, state}
  end

  def handle_call({:prompt, text}, from, %{active_from: nil} = state) do
    {:noreply, start_prompt(state, text, from)}
  end

  def handle_call({:prompt, text}, from, state) do
    # A prior turn on this session is still in flight — one ACP session is
    # one conversation, so this turn is queued rather than interleaved.
    {:noreply, %{state | queue: :queue.in({text, from}, state.queue)}}
  end

  def handle_call(:close, _from, state) do
    {:stop, :normal, :ok, state}
  end

  defp start_prompt(state, text, from) do
    id = "p-#{state.next_id}"

    Stdio.send_msg(state.stdio, %{
      "jsonrpc" => "2.0",
      "id" => id,
      "method" => "session/prompt",
      "params" => %{
        "sessionId" => state.session_id,
        "prompt" => [%{"type" => "text", "text" => text}]
      }
    })

    %{state | next_id: state.next_id + 1, active_from: from, active_id: id, chunks: []}
  end

  # -- inbound routing ----------------------------------------------------------

  @impl true
  def handle_info({port, {:data, chunk}}, %{stdio: %Stdio{port: port}} = state) do
    {lines, rest} = Stdio.split_lines(state.buffer <> chunk)
    state = %{state | buffer: rest}
    state = Enum.reduce(lines, state, &route_line/2)
    {:noreply, state}
  end

  def handle_info({port, {:exit_status, _code}}, %{stdio: %Stdio{port: port}} = state) do
    state = fail_all(state, :acp_process_exited)
    {:noreply, %{state | closed: true}}
  end

  def handle_info(_other, state), do: {:noreply, state}

  defp route_line(line, state) do
    case Jason.decode(line) do
      {:ok, msg} when is_map(msg) -> route(msg, state)
      _ -> state
    end
  end

  defp route(%{"id" => id, "result" => _}, %{active_id: id} = state) do
    finish_active(state, {:ok, join_chunks(state)})
  end

  defp route(%{"id" => id, "error" => err}, %{active_id: id} = state) do
    finish_active(state, {:error, {:acp_error, err}})
  end

  defp route(%{"method" => "session/update", "params" => params}, state) do
    accumulate(state, params)
  end

  defp route(%{"method" => "session/request_permission", "id" => id, "params" => params}, state) do
    answer_permission(state.stdio, id, params)
    state
  end

  defp route(_msg, state), do: state

  defp accumulate(state, %{"update" => %{"sessionUpdate" => "agent_message_chunk"} = update}) do
    text = get_in(update, ["content", "text"]) || ""
    %{state | chunks: [text | state.chunks]}
  end

  defp accumulate(state, _params), do: state

  defp answer_permission(stdio, id, params) do
    options = params["options"] || []

    chosen =
      Enum.find_value(options, fn o ->
        kind = to_string(o["kind"] || "")
        if String.starts_with?(kind, "allow"), do: o["optionId"]
      end)

    result =
      if chosen do
        %{"outcome" => %{"outcome" => "selected", "optionId" => chosen}}
      else
        %{"outcome" => %{"outcome" => "cancelled"}}
      end

    Stdio.send_msg(stdio, %{"jsonrpc" => "2.0", "id" => id, "result" => result})
  end

  defp join_chunks(state), do: state.chunks |> Enum.reverse() |> Enum.join("")

  defp finish_active(state, outcome) do
    if state.active_from, do: GenServer.reply(state.active_from, outcome)
    state = %{state | active_from: nil, active_id: nil, chunks: []}
    dequeue_next(state)
  end

  defp dequeue_next(state) do
    case :queue.out(state.queue) do
      {{:value, {text, from}}, rest} -> start_prompt(%{state | queue: rest}, text, from)
      {:empty, _} -> state
    end
  end

  defp fail_all(state, reason) do
    if state.active_from, do: GenServer.reply(state.active_from, {:error, reason})

    state.queue
    |> :queue.to_list()
    |> Enum.each(fn {_text, from} -> GenServer.reply(from, {:error, reason}) end)

    %{state | active_from: nil, active_id: nil, queue: :queue.new()}
  end

  @impl true
  def terminate(_reason, state) do
    if state[:stdio], do: Stdio.close(state.stdio)
    :ok
  end
end
